<a href="https://colab.research.google.com/github/MYon003/Myon/blob/main/Labs/Michael_pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Intro to Pandas

![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/pandas-logo-2.png)

<font size='0.85em'>(logo by [Marc Garcia](https://github.com/pandas-dev/pandas/blob/master/web/pandas/static/img/pandas.svg))</font> 


> A suggestion: This worksheet is organized in a way that, hopefully, best presents the topics and not in a way that is the most useful for reference. As you progress through this worksheet, your future you will find it useful if you create your own notes so that you can quickly find the relevant information. These notes might take the form of a separate Google Colab notebook or handwritten notes. In addition, just the activity of taking notes will help you retain the information.

The Pandas Library is built on top of Numpy and is designed to make working with data fast and easy. Like Numpy, the library includes data structures and functions to manipulate that data.

As we learned in the Numpy Notebook, we need to load in the library before we can use it.

In [86]:
from pandas import Series, DataFrame
import numpy as np
import pandas as pd

Let's dissect the code above.

The `Series` and `DataFrame` datatypes are commonly used so we import them directly with

```
from pandas import Series, DataFrame
```

For all other datatypes and functions in the library, the `pd` prefix is commonly used so we import that with

```
import pandas as pd
```

## Series
A series is a 1d array-like object

Let's consider the heights (in cm) of the members of the Japan's Women's Basketball Team at the 2020 Olympics (they won the silver medal).      

![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/basketball2020.png)
We can create a series in a number of ways.

For example, we can create it directly from a Python list

In [87]:
japan = [183, 185, 167, 162, 165, 174, 173, 181, 167, 183, 185, 182] # a Python list
athletesHeight = Series(japan) # converted to a Pandas Series

We could also have done

In [88]:
athletesHeight = Series([183, 185, 167, 162, 165, 174, 173, 181, 167, 183, 185, 182])

In either case we can see the value of the Series `athletesHeight`

In [89]:
 athletesHeight

0     183
1     185
2     167
3     162
4     165
5     174
6     173
7     181
8     167
9     183
10    185
11    182
dtype: int64

You are probably familiar with arrays in programming languages and this works in a similar way. The left number is the index and the right the value. And we can find the value at a particular index by the usual:

In [90]:
athletesHeight[3]

162

#### specifying indices
Instead of the index 0, 1, 2, 3 ... you can specify your own index values. For example, we can label them 'Moeko Nagaoka', 'Maki Takada', 'Naho Miyoshi' ... etc. 

In [91]:
athletes2 = Series(japan, index = ['Moeko Nagaoka', 'Maki Takada', 'Naho Miyoshi', 
                                   'Rui Machida', 'Nako Motohashi', 'Nanaka Todo', 
                                   'Saki Hayashi','Evelyn Mawuli', 'Saori Miyazaki', 'Yuki Miyazawa', 'Himawari Akaho', 
                                   'Monica Okoye'])
athletes2

Moeko Nagaoka     183
Maki Takada       185
Naho Miyoshi      167
Rui Machida       162
Nako Motohashi    165
Nanaka Todo       174
Saki Hayashi      173
Evelyn Mawuli     181
Saori Miyazaki    167
Yuki Miyazawa     183
Himawari Akaho    185
Monica Okoye      182
dtype: int64

The names we see are not another column of the data. We can see the shape of athletes2 by:

In [92]:
athletes2.shape

(12,)

This shows that athletes2 is a one dimensional matrix and that dimension has a length of 12. So the names we see are not values in a column but rather the indices.


Let's use the index to get the Height of Himawari Akaho:

In [93]:
athletes2['Himawari Akaho']

185


![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/divider.png)
## DataFrame
DataFrames are **the most important data structure of Pandas** and are simply
a table or spreadsheet like structure.  A DataFrame represents a table like:
![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/ioniq5s.png)

Make | Drive | Range | Battery_kwH | PeakPower_kW | MPGe | 0-60
:---- | :--- | ---: | ---: | ---: | ---: | ---:
Ioniq 5 | AWD | 256 | 77.4 | 239 | 98 | 5.0
Tesla Model 3 | AWD | 315 | 80 | 298 | 113 | 3.1
Rivian R1T | AWD | 316 | 135 | 562 | 70 | 3.0 
VW ID.4 | RWD | 260 | 82 | 260 | 107 |  7.7
F150 Lightning | AWD | 230 | 110 | 318 | 68 | 4.0

### Creating a DataFrame by Hand

A common way to create a DataFrame is to use a Python dictionary as follows:

In [94]:
 cars = {'Make': ['Ioniq 5', 'Tesla Model 3', 'Rivian R1T', 'VW ID.4', 'F150 Lightning'],
         'Drive': ['AWD', 'AWD', 'AWD', 'RWD', 'AWD'],
         'Range': [256, 315, 316, 260, 230],
         'Battery_kwH': [77.4, 80, 135, 82, 110],
         'PeakPower_kW': [239, 298, 562, 260, 318],
         'MPGe': [98, 113, 70, 107, 68],
         '0-60': [5.0, 3.1, 3.0, 7.7, 4.0]}


and now we can create a DataFrame from the `cars` Python dictionary:

In [95]:
df = DataFrame(cars)
df


Make Drive  Range  Battery_kwH  PeakPower_kW  MPGe  0-60
0         Ioniq 5   AWD    256         77.4           239    98   5.0
1   Tesla Model 3   AWD    315         80.0           298   113   3.1
2      Rivian R1T   AWD    316        135.0           562    70   3.0
3         VW ID.4   RWD    260         82.0           260   107   7.7
4  F150 Lightning   AWD    230        110.0           318    68   4.0

Prior to my life with Pandas, I would represent a table like the above one as:

In [96]:
prePandas = [{'Make': 'Ioniq 5', 'Drive': 'AWD', 'Range': 256, 'Battery_kwH': 77.4, 'PeakPower_kW': 98, 'MPGe': 98, '0-60': 5.0},
             {'Make': 'Tesla Model 3', 'Drive': 'AWD', 'Range': 315, 'Battery_kwH': 80.0, 'PeakPower_kW': 113,'MPGe': 113, '0-60': 3.1},
             {'Make': 'Rivian R1T', 'Drive': 'AWD', 'Range': 316, 'Battery_kwH': 135.0, 'PeakPower_kW': 562,'MPGe': 70, '0-60': 3.0},
             {'Make': 'VW ID.4', 'Drive': 'RWD', 'Range': 260, 'Battery_kwH': 82.0, 'PeakPower_kW': 260,'MPGe': 107, '0-60': 7.7},
             {'Make': 'F150 Lightning', 'Drive': 'AWD', 'Range': 230, 'Battery_kwH': 110.0, 'PeakPower_kW': 318,'MPGe': 68, '0-60': 4.0}]


prePandas[0]['Make']

'Ioniq 5'

In the prePandas scheme the data is organized first by rows. That seemed logical to me since each row represents an object and is how we organize data in an SQL database. In the Pandas representation the data is organized by columns. 



![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/divider.png)

## Reading data from a csv file:

A CSV file (comma separated values) is a common data format for data science files. As| the name suggests, the data values in a row are separated by commas. For example, the data file to represent the table:


Make | Drive | Range | Battery_kwH | PeakPower_kW | MPGe | 0-60
:---- | :--- | ---: | ---: | ---: | ---: | ---:
Ioniq 5 | AWD | 256 | 77.4 | 239 | 98 | 5.0
Tesla Model 3 | AWD | 315 | 80 | 298 | 113 | 3.1
Rivian R1T | AWD | 316 | 135 | 562 | 70 | 3.0 
VW ID.4 | RWD | 260 | 82 | 260 | 107 |  7.7
F150 Lightning | AWD | 230 | 110 | 318 | 68 | 4.0


would be

```
Make,Drive,Range,Battery_kwH,PeakPower_kW,MPGe,0-60
Ioniq 5,AWD,256,77.4,239,98,5.0
Tesla Model 3,AWD,315,80,298,113,3.1
Rivian R1T,AWD,316,135,562,70,3.0 
VW ID.4,RWD,260,82,260,107,7.7
F150 Lightning,AWD,230,110,318,68,4.0

```


As the name suggests, we use the `pd.read_csv` function to read a csv file.  `pd.read_csv` can read a csv file from either your local machine or the web. Let's start with the web.

### Reading a CSV file from the web.
To read a file from the web, we simply provide a URL:


In [97]:
evs = pd.read_csv('https://raw.githubusercontent.com/zacharski/datamining-guide/main/data/ElectricCarData_Clean.csv')
evs

Brand                            Model  AccelSec  TopSpeed_KmH  \
0         Tesla    Model 3 Long Range Dual Motor       4.6           233   
1    Volkswagen                        ID.3 Pure      10.0           160   
2      Polestar                                2       4.7           210   
3           BMW                              iX3       6.8           180   
4         Honda                                e       9.5           145   
..          ...                              ...       ...           ...   
101      Nissan  Ariya e-4ORCE 87kWh Performance       5.1           200   
102       Byton                M-Byte 95 kWh 2WD       7.5           190   
103      Rivian                              R1T       3.0           185   
104         GMC                        Hummer EV       3.0           171   
105        Ford                   F150 Lightning       3.0           180   

     Range_Km  Efficiency_WhKm FastCharge_KmH RapidCharge PowerTrain  \
0         450              161            940         Yes        AWD   
1         270              167            250         Yes        RWD   
2         400              181            620         Yes        AWD   
3         360              206            560         Yes        RWD   
4         170              168            190         Yes        RWD   
..        ...              ...            ...         ...        ...   
101       375              232            450         Yes        AWD   
102       400              238            480         Yes        AWD   
103       505              299            337         Yes        AWD   
104       529              446           250          Yes        AWD   
105       514              299           309          Yes        AWD   

       PlugType  BodyStyle Segment  Seats PriceEuro  
0    Type 2 CCS      Sedan       D      5     55480  
1    Type 2 CCS  Hatchback       C      5     30000  
2    Type 2 CCS   Liftback       D      5     56440  
3    Type 2 CCS        SUV       D      5     68040  
4    Type 2 CCS  Hatchback       B      4     32997  
..          ...        ...     ...    ...       ...  
101  Type 2 CCS  Hatchback       C      5     65000  
102  Type 2 CCS        SUV       E      5     62000  
103  Type 2 CCS     Pickup       N      6     75000  
104  Type 2 CCS      Truck       E      5         -  
105  Type 2 CCS      Truck       E      5         -  

[106 rows x 14 columns]

Sometimes the csv file has a header row as was the case in the example above. That file starts with the line
```
Brand,Model,AccelSec,TopSpeed_KmH,Range_Km,Efficiency_WhKm,FastCharge_KmH,RapidCharge,PowerTrain,PlugType,BodyStyle,Segment,Seats,PriceEuro
```
as we can see by using `curl`



In [98]:
!curl -s https://raw.githubusercontent.com/zacharski/datamining-guide/main/data/ElectricCarData_Clean.csv |  head

Brand,Model,AccelSec,TopSpeed_KmH,Range_Km,Efficiency_WhKm,FastCharge_KmH,RapidCharge,PowerTrain,PlugType,BodyStyle,Segment,Seats,PriceEuro
Tesla,Model 3 Long Range Dual Motor,4.6,233,450,161,940,Yes,AWD,Type 2 CCS,Sedan,D,5,55480
Volkswagen,ID.3 Pure,10,160,270,167,250,Yes,RWD,Type 2 CCS,Hatchback,C,5,30000
Polestar,2,4.7,210,400,181,620,Yes,AWD,Type 2 CCS,Liftback,D,5,56440
BMW,iX3,6.8,180,360,206,560,Yes,RWD,Type 2 CCS,SUV,D,5,68040
Honda,e,9.5,145,170,168,190,Yes,RWD,Type 2 CCS,Hatchback,B,4,32997
Lucid,Air,2.8,250,610,180,620,Yes,AWD,Type 2 CCS,Sedan,F,5,105000
Volkswagen,e-Golf,9.6,150,190,168,220,Yes,FWD,Type 2 CCS,Hatchback,C,5,31900
Peugeot,e-208,8.1,150,275,164,420,Yes,FWD,Type 2 CCS,Hatchback,B,5,29682
Tesla,Model 3 Standard Range Plus,5.6,225,310,153,650,Yes,RWD,Type 2 CCS,Sedan,D,5,46380


**An aside**

We can preface any Unix command with a bang (!) to have it execute in our Notebook.  This is amazingly handy:



In [99]:
!ls

8.csv  data  sample_data  tryit.csv


In [100]:
!pwd

/content


### Files with No Header Row
Sometimes the csv file does not have a header row. So for example, data might start on the very first line of the file

```
Tesla ,Model 3 Long Range Dual Motor,4.6,233,450,161,940,Yes,AWD,Type 2 CCS,Sedan,D,5,55480
Volkswagen ,ID.3 Pure,10,160,270,167,250,Yes,RWD,Type 2 CCS,Hatchback,C,5,30000
Polestar ,2,4.7,210,400,181,620,Yes,AWD,Type 2 CCS,Liftback,D,5,56440
```

In that case you specify the names of the columns using the `names` parameter:

In [101]:
columnNames = ['Brand','Model','AccelSec','TopSpeed_KmH','Range_Km','Efficiency_WhKm',
               'FastCharge_KmH','RapidCharge','PowerTrain','PlugType','BodyStyle',
               'Segment','Seats','PriceEuro']

evs2 = pd.read_csv('https://raw.githubusercontent.com/zacharski/datamining-guide/main/data/ElectricCarData_Clean_NoHeader.csv', names=columnNames)
evs2

Brand                            Model  AccelSec  TopSpeed_KmH  \
0         Tesla     Model 3 Long Range Dual Motor       4.6           233   
1    Volkswagen                         ID.3 Pure      10.0           160   
2      Polestar                                 2       4.7           210   
3           BMW                              iX3        6.8           180   
4         Honda                                e        9.5           145   
..           ...                              ...       ...           ...   
98       Nissan                       Ariya 63kWh       7.5           160   
99         Audi     e-tron S Sportback 55 quattro       4.5           210   
100      Nissan               Ariya e-4ORCE 63kWh       5.9           200   
101      Nissan   Ariya e-4ORCE 87kWh Performance       5.1           200   
102       Byton                 M-Byte 95 kWh 2WD       7.5           190   

     Range_Km  Efficiency_WhKm FastCharge_KmH RapidCharge PowerTrain  \
0         450              161            940         Yes        AWD   
1         270              167            250         Yes        RWD   
2         400              181            620         Yes        AWD   
3         360              206            560         Yes        RWD   
4         170              168            190         Yes        RWD   
..        ...              ...            ...         ...        ...   
98        330              191            440         Yes        FWD   
99        335              258            540         Yes        AWD   
100       325              194            440         Yes        AWD   
101       375              232            450         Yes        AWD   
102       400              238            480         Yes        AWD   

       PlugType  BodyStyle Segment  Seats  PriceEuro  
0    Type 2 CCS      Sedan       D      5      55480  
1    Type 2 CCS  Hatchback       C      5      30000  
2    Type 2 CCS   Liftback       D      5      56440  
3    Type 2 CCS        SUV       D      5      68040  
4    Type 2 CCS  Hatchback       B      4      32997  
..          ...        ...     ...    ...        ...  
98   Type 2 CCS  Hatchback       C      5      45000  
99   Type 2 CCS        SUV       E      5      96050  
100  Type 2 CCS  Hatchback       C      5      50000  
101  Type 2 CCS  Hatchback       C      5      65000  
102  Type 2 CCS        SUV       E      5      62000  

[103 rows x 14 columns]


![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/divider.png)

### Reading a CSV file the local machine
First, let's get that file onto our local machine:

In [102]:
!curl https://raw.githubusercontent.com/zacharski/datamining-guide/main/data/ElectricCarData_Clean.csv > evs.csv

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  8268  100  8268    0     0   207k      0 --:--:-- --:--:-- --:--:--  212k


Hmm. That didn't work. Can you fix the error and rerun that cell?

---
Now we can specify the local file using `pd=read_csv`

In [103]:
ev3 = pd.read_csv('evs.csv')
ev3

Brand                            Model  AccelSec  TopSpeed_KmH  \
0         Tesla    Model 3 Long Range Dual Motor       4.6           233   
1    Volkswagen                        ID.3 Pure      10.0           160   
2      Polestar                                2       4.7           210   
3           BMW                              iX3       6.8           180   
4         Honda                                e       9.5           145   
..          ...                              ...       ...           ...   
101      Nissan  Ariya e-4ORCE 87kWh Performance       5.1           200   
102       Byton                M-Byte 95 kWh 2WD       7.5           190   
103      Rivian                              R1T       3.0           185   
104         GMC                        Hummer EV       3.0           171   
105        Ford                   F150 Lightning       3.0           180   

     Range_Km  Efficiency_WhKm FastCharge_KmH RapidCharge PowerTrain  \
0         450              161            940         Yes        AWD   
1         270              167            250         Yes        RWD   
2         400              181            620         Yes        AWD   
3         360              206            560         Yes        RWD   
4         170              168            190         Yes        RWD   
..        ...              ...            ...         ...        ...   
101       375              232            450         Yes        AWD   
102       400              238            480         Yes        AWD   
103       505              299            337         Yes        AWD   
104       529              446           250          Yes        AWD   
105       514              299           309          Yes        AWD   

       PlugType  BodyStyle Segment  Seats PriceEuro  
0    Type 2 CCS      Sedan       D      5     55480  
1    Type 2 CCS  Hatchback       C      5     30000  
2    Type 2 CCS   Liftback       D      5     56440  
3    Type 2 CCS        SUV       D      5     68040  
4    Type 2 CCS  Hatchback       B      4     32997  
..          ...        ...     ...    ...       ...  
101  Type 2 CCS  Hatchback       C      5     65000  
102  Type 2 CCS        SUV       E      5     62000  
103  Type 2 CCS     Pickup       N      6     75000  
104  Type 2 CCS      Truck       E      5         -  
105  Type 2 CCS      Truck       E      5         -  

[106 rows x 14 columns]

Suppose we want that file in a data directory. Let's go ahead and create the directory and move the file there.

In [104]:
!mkdir data
!mv evs.csv data

mkdir: cannot create directory ‘data’: File exists


Now when we load the file we need to give more of a path:

In [105]:
evs4 = pd.read_csv('data/evs.csv')
evs4

Brand                            Model  AccelSec  TopSpeed_KmH  \
0         Tesla    Model 3 Long Range Dual Motor       4.6           233   
1    Volkswagen                        ID.3 Pure      10.0           160   
2      Polestar                                2       4.7           210   
3           BMW                              iX3       6.8           180   
4         Honda                                e       9.5           145   
..          ...                              ...       ...           ...   
101      Nissan  Ariya e-4ORCE 87kWh Performance       5.1           200   
102       Byton                M-Byte 95 kWh 2WD       7.5           190   
103      Rivian                              R1T       3.0           185   
104         GMC                        Hummer EV       3.0           171   
105        Ford                   F150 Lightning       3.0           180   

     Range_Km  Efficiency_WhKm FastCharge_KmH RapidCharge PowerTrain  \
0         450              161            940         Yes        AWD   
1         270              167            250         Yes        RWD   
2         400              181            620         Yes        AWD   
3         360              206            560         Yes        RWD   
4         170              168            190         Yes        RWD   
..        ...              ...            ...         ...        ...   
101       375              232            450         Yes        AWD   
102       400              238            480         Yes        AWD   
103       505              299            337         Yes        AWD   
104       529              446           250          Yes        AWD   
105       514              299           309          Yes        AWD   

       PlugType  BodyStyle Segment  Seats PriceEuro  
0    Type 2 CCS      Sedan       D      5     55480  
1    Type 2 CCS  Hatchback       C      5     30000  
2    Type 2 CCS   Liftback       D      5     56440  
3    Type 2 CCS        SUV       D      5     68040  
4    Type 2 CCS  Hatchback       B      4     32997  
..          ...        ...     ...    ...       ...  
101  Type 2 CCS  Hatchback       C      5     65000  
102  Type 2 CCS        SUV       E      5     62000  
103  Type 2 CCS     Pickup       N      6     75000  
104  Type 2 CCS      Truck       E      5         -  
105  Type 2 CCS      Truck       E      5         -  

[106 rows x 14 columns]


![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/divider.png)

## Missing Data


In real machine learning tasks, we often encounter missing values.

### Missing Data in Files
For example, suppose we didn't know The Polestar 2's acceleration, the BMW iX3's top speed and the Honda's model name. In that case our CSV file would start


```
Brand,Model,AccelSec,TopSpeed_KmH,Range_Km
Tesla ,Model 3 Long Range Dual Motor,4.6,233,450
Volkswagen ,ID.3 Pure,10.0,160,270
Polestar ,2,,210,400
BMW ,iX3 ,6.8,,360
Honda ,,9.5,145,
```

with the double comma on the Polestar, BMW and Honda lines representing the missing data. When we read that file.



In [106]:
evs5 = pd.read_csv('https://raw.githubusercontent.com/zacharski/datamining-guide/main/data/ev_cars_small.csv')
evs5

Brand                          Model  AccelSec  TopSpeed_KmH  \
0        Tesla   Model 3 Long Range Dual Motor       4.6         233.0   
1   Volkswagen                       ID.3 Pure      10.0         160.0   
2     Polestar                               2       NaN         210.0   
3          BMW                            iX3        6.8           NaN   
4        Honda                             NaN       9.5         145.0   
5        Lucid                            Air        2.8         250.0   
6   Volkswagen                         e-Golf        9.6         150.0   
7      Peugeot                          e-208        8.1         150.0   
8        Tesla     Model 3 Standard Range Plus       5.6         225.0   
9         Audi                      Q4 e-tron        6.3         180.0   
10    Mercedes                  EQC 400 4MATIC       5.1         180.0   
11      Nissan                           Leaf        7.9         144.0   
12     Hyundai            Kona Electric 64 kWh       7.9         167.0   
13         BMW                             i4        4.0         200.0   
14     Hyundai                  IONIQ Electric       9.7         165.0   
15  Volkswagen                      ID.3 Pro S       7.9         160.0   
16     Porsche                  Taycan Turbo S       2.8         260.0   
17  Volkswagen                          e-Up!       11.9         130.0   
18          MG                           ZS EV       8.2         140.0   
19        Mini                      Cooper SE        7.3         150.0   

    Range_Km  
0      450.0  
1      270.0  
2      400.0  
3      360.0  
4        NaN  
5      610.0  
6      190.0  
7      275.0  
8      310.0  
9      400.0  
10     370.0  
11     220.0  
12     400.0  
13     450.0  
14     250.0  
15     440.0  
16     375.0  
17     195.0  
18     220.0  
19     185.0

We see that missing values are now the floating point values NaN meaning Not a Number. This NaN value is used even in columns that do not contain floating point values. For example, in row 4 above, Honda has NaN in the model name column. Wes McKinney, the developer of Pandas calls NaN a *sentinel* value that is easily detected and indicates a missing value.

### Special Characters Representing NaN in the Data File

Sometimes, special characters are used in a data file to represent missing values. For example, sometimes a dash is used ...

Brand|Model|AccelSec|TopSpeed_KmH|Range_Km
:--- | :--- | ---: | ---: | ---: |
Tesla |Model 3 Long Range Dual Motor|4.6|233|450
Volkswagen |ID.3 Pure|10.0|160|270
Polestar |2|-|210|400
BMW |iX3 |6.8|-|360
Honda |-|9.5|145|

The associated data file would look like

```
Brand,Model,AccelSec,TopSpeed_KmH,Range_Km
Tesla ,Model 3 Long Range Dual Motor,4.6,233,450
Volkswagen ,ID.3 Pure,10.0,160,270
Polestar ,2,-,210,400
BMW ,iX3 ,6.8,-,360
Honda ,-,9.5,145,
```

If we do not convert these to `NaN`, these dashes will create havoc with future calculations:

``` 
TypeError: can't multiply sequence by non-int of type 'float'
```

When we read in the csv file we need to convert the dashes to `NaN` by using the `na_values` parameter in `read_csv`. As you can see in the following example, our data file has the dashes 

In [107]:
!curl https://raw.githubusercontent.com/zacharski/datamining-guide/main/data/ev_cars_small-dash.csv

Brand,Model,AccelSec,TopSpeed_KmH,Range_Km
Tesla ,Model 3 Long Range Dual Motor,4.6,233,450
Volkswagen ,ID.3 Pure,10.0,160,270
Polestar ,2,-,210,400
BMW ,iX3 ,6.8,-,360
Honda ,-,9.5,145,
Lucid ,Air ,2.8,250,610
Volkswagen ,e-Golf ,9.6,150,190
Peugeot ,e-208 ,8.1,150,275
Tesla ,Model 3 Standard Range Plus,5.6,225,310
Audi ,Q4 e-tron ,6.3,180,400
Mercedes ,EQC 400 4MATIC,5.1,180,370
Nissan ,Leaf ,7.9,144,220
Hyundai ,Kona Electric 64 kWh,7.9,167,400
BMW ,i4 ,4.0,200,450
Hyundai ,IONIQ Electric,9.7,165,250
Volkswagen ,ID.3 Pro S,7.9,160,440
Porsche ,Taycan Turbo S,2.8,260,375
Volkswagen ,e-Up! ,11.9,130,195
MG ,ZS EV,8.2,140,220
Mini ,Cooper SE ,7.3,150,185


and we can convert those to `NaN` when we read the file

In [108]:
evdash = pd.read_csv('https://raw.githubusercontent.com/zacharski/datamining-guide/main/data/ev_cars_small-dash.csv',
                     na_values='-')
evdash[:6]

Brand                          Model  AccelSec  TopSpeed_KmH  \
0       Tesla   Model 3 Long Range Dual Motor       4.6         233.0   
1  Volkswagen                       ID.3 Pure      10.0         160.0   
2    Polestar                               2       NaN         210.0   
3         BMW                            iX3        6.8           NaN   
4       Honda                             NaN       9.5         145.0   
5       Lucid                            Air        2.8         250.0   

   Range_Km  
0     450.0  
1     270.0  
2     400.0  
3     360.0  
4       NaN  
5     610.0

### Specifying Missing Values by hand
Suppose we didn't know the range of the Hyundai Ioniq 5 and the MPGe of the Tesla.

![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/ioniq5s.png)

Make | Drive | Range | Battery_kwH | PeakPower_kW | MPGe | 0-60
:---- | :--- | ---: | ---: | ---: | ---: | ---:
Ioniq 5 | AWD | - | 77.4 | 239 | 98 | 5.0
Tesla Model 3 | AWD | 315 | 80 | 298 | - | 3.1
Rivian R1T | AWD | 316 | 135 | 562 | 70 | 3.0 
VW ID.4 | RWD | 260 | 82 | 260 | 107 |  7.7
F150 Lightning | AWD | 230 | 110 | 318 | 68 | 4.0

In that case we can create a dataframe like:

In [109]:
cars3 = {'Make': ['Ioniq 5', 'Tesla Model 3', 'Rivian R1T', 'VW ID.4', 'F150 Lightning'],
         'Drive': ['AWD', 'AWD', 'AWD', 'RWD', 'AWD'],
         'Range': [np.nan, 315, 316, 260, 230],
         'Battery_kwH': [77.4, 80, 135, 82, 110],
         'PeakPower_kW': [239, 298, 562, 260, 318],
         'MPGe': [98, np.nan, 70, 107, 68],
         '0-60': [5.0, 3.1, 3.0, 7.7, 4.0]}
        
carz = DataFrame(cars3)
carz

Make Drive  Range  Battery_kwH  PeakPower_kW   MPGe  0-60
0         Ioniq 5   AWD    NaN         77.4           239   98.0   5.0
1   Tesla Model 3   AWD  315.0         80.0           298    NaN   3.1
2      Rivian R1T   AWD  316.0        135.0           562   70.0   3.0
3         VW ID.4   RWD  260.0         82.0           260  107.0   7.7
4  F150 Lightning   AWD  230.0        110.0           318   68.0   4.0

where `np.nan` is Numpy's NaN.  We can also use Python's `None`:


In [110]:
cars3 = {'Make': ['Ioniq 5', 'Tesla Model 3', 'Rivian R1T', 'VW ID.4', 'F150 Lightning'],
         'Drive': ['AWD', 'AWD', 'AWD', 'RWD', 'AWD'],
         'Range': [None, 315, 316, 260, 230],
         'Battery_kwH': [77.4, 80, 135, 82, 110],
         'PeakPower_kW': [239, 298, 562, 260, 318],
         'MPGe': [98, None, 70, 107, 68],
         '0-60': [5.0, 3.1, 3.0, 7.7, 4.0]}
        
carz = DataFrame(cars3)
carz

Make Drive  Range  Battery_kwH  PeakPower_kW   MPGe  0-60
0         Ioniq 5   AWD    NaN         77.4           239   98.0   5.0
1   Tesla Model 3   AWD  315.0         80.0           298    NaN   3.1
2      Rivian R1T   AWD  316.0        135.0           562   70.0   3.0
3         VW ID.4   RWD  260.0         82.0           260  107.0   7.7
4  F150 Lightning   AWD  230.0        110.0           318   68.0   4.0

In addition to reading CSV files, there are many other ways of reading in data including from SQL databases, mongoDB, and webpages. See the [Pandas documentation](https://pandas.pydata.org/pandas-docs/stable/) for details.



![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/divider.png)

## Indices
In all the examples above, when we created a DataFrame, an auto-generated, auto-incrementing index was generated. So our initial data might look like:


Make | Drive | Range | Battery_kwH | PeakPower_kW | MPGe | 0-60
:---- | :--- | ---: | ---: | ---: | ---: | ---:
Ioniq 5 | AWD | - | 77.4 | 239 | 98 | 5.0
Tesla Model 3 | AWD | 315 | 80 | 298 | - | 3.1
Rivian R1T | AWD | 316 | 135 | 562 | 70 | 3.0 
VW ID.4 | RWD | 260 | 82 | 260 | 107 |  7.7
F150 Lightning | AWD | 230 | 110 | 318 | 68 | 4.0

But the DataFrame looks like ...

In [111]:
carz
   

Make Drive  Range  Battery_kwH  PeakPower_kW   MPGe  0-60
0         Ioniq 5   AWD    NaN         77.4           239   98.0   5.0
1   Tesla Model 3   AWD  315.0         80.0           298    NaN   3.1
2      Rivian R1T   AWD  316.0        135.0           562   70.0   3.0
3         VW ID.4   RWD  260.0         82.0           260  107.0   7.7
4  F150 Lightning   AWD  230.0        110.0           318   68.0   4.0

with the 0, 1, 2 ... indices added (the column on the left). This makes it handy when we want to access a particular row.

In [112]:
carz.loc[4]

Make            F150 Lightning
Drive                      AWD
Range                    230.0
Battery_kwH              110.0
PeakPower_kW               318
MPGe                      68.0
0-60                       4.0
Name: 4, dtype: object

As you can see in the above code, we access a particular row (or rows) with the `loc` method.

Sometimes the rows in our data already have a unique identifier. For example in a data file for U.S. states each state might have the unique 2 character state code


![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/grayStates.png)

State|Name|Pop2022|Pop2021|Pop2010|TotalArea|LandArea
:--- | :--- | ---: | ---: | ---: | ---: | ---: | 
AK| Alaska| 720763|724357|713910|665384|570640.95
AL| Alabama|4949697|4934193|4785437|52420|50645.33
AR| Arkansas|3042017|3033946|2921964|53179|52035.48
AZ| Arizona|7640796|7520103|6407172|113990|113594.08

In that case we can use the unique identifier, in this case the 2 letter state code, as our index. 

In [113]:
states = pd.read_csv('https://raw.githubusercontent.com/zacharski/datamining-guide/main/data/usStates.csv',
                     index_col='State')
states

Name   Pop2022   Pop2021   Pop2010  TotalArea   LandArea
State                                                                    
AK             Alaska    720763    724357    713910     665384  570640.95
AL            Alabama   4949697   4934193   4785437      52420   50645.33
AR           Arkansas   3042017   3033946   2921964      53179   52035.48
AZ            Arizona   7640796   7520103   6407172     113990  113594.08
CA         California  39664128  39613493  37319502     163695  155779.22
CO           Colorado   5961083   5893634   5047349     104094  103641.89
CT        Connecticut   3546588   3552821   3579114       5543    4842.36
DE           Delaware    998619    990334    899593       2489    1948.54
FL            Florida  22177997  21944577  18845537      65758   53624.76
GA            Georgia  10936299  10830007   9711881      59425   57513.49
HI             Hawaii   1401709   1406430   1363963      10932    6422.63
IA               Iowa   3174426   3167974   3050745      56273   55857.13
ID              Idaho   1896652   1860123   1570746      83569   82643.12
IL           Illinois  12518071  12569321  12840503      57914   55518.93
IN            Indiana   6842385   6805663   6490432      36420   35826.11
KS             Kansas   2919179   2917224   2858190      82278   81758.72
KY           Kentucky   4487233   4480713   4348181      40408   39486.34
LA          Louisiana   4616106   4627002   4544532      52378   43293.90
MA      Massachusetts   6922107   6912239   6566307      10554    7800.06
MD           Maryland   6075314   6065436   5788645      12406    9707.24
ME              Maine   1359677   1354522   1327629      35380   30842.92
MI           Michigan   9995212   9992427   9877510      96714   56538.90
MN          Minnesota   5739781   5706398   5310828      86936   79626.74
MO           Missouri   6184843   6169038   5995974      69707   68741.52
MS        Mississippi   2961536   2966407   2970548      48432   43923.27
MT            Montana   1093117   1085004    990697     147040  145545.80
NC     North Carolina  10807491  10701022   9574323      53819   48617.91
ND       North Dakota    774008    770026    674715      70698   69000.80
NE           Nebraska   1960790   1951996   1829542      77348   76824.17
NH      New Hampshire   1378449   1372203   1316762       9349    8952.65
NJ         New Jersey   8870685   8874520   8799446       8723    7354.22
NM         New Mexico   2109093   2105005   2064552     121590  121298.15
NV             Nevada   3238601   3185786   2702405     110572  109781.18
NY           New York  19223191  19299981  19399878      54555   47126.40
OH               Ohio  11727377  11714618  11539336      44826   40860.69
OK           Oklahoma   4007179   3990443   3759944      69899   68594.92
OR             Oregon   4325290   4289439   3837491      98379   95988.01
PA       Pennsylvania  12805190  12804123  12711160      46054   44742.70
RI       Rhode Island   1062583   1061509   1053959       1545    1033.81
SC     South Carolina   5342388   5277830   4635649      32020   30060.70
SD       South Dakota    902542    896581    816166      77116   75811.00
TN          Tennessee   7001803   6944260   6355311      42114   41234.90
TX              Texas  30097526  29730311  25241971     268596  261231.71
UT               Utah   3363182   3310774   2775332      84897   82169.62
VA           Virginia   8638218   8603985   8023699      42775   39490.09
VT            Vermont    622882    623251    625879       9616    9216.66
WA         Washington   7887965   7796941   6742830      71298   66455.52
WI          Wisconsin   5867518   5852490   5690475      65496   54157.80
WV      West Virginia   1755715   1767859   1854239      24230   24038.21
WY            Wyoming    582233    581075    564487      97813   97093.14

We can access a particular row, say the info for New Mexico by 

In [114]:
states.loc['NM']

Name         New Mexico
Pop2022         2109093
Pop2021         2105005
Pop2010         2064552
TotalArea        121590
LandArea      121298.15
Name: NM, dtype: object

You can access a range of rows by the standard Python method:

In [115]:
states[:5]

Name   Pop2022   Pop2021   Pop2010  TotalArea   LandArea
State                                                                
AK         Alaska    720763    724357    713910     665384  570640.95
AL        Alabama   4949697   4934193   4785437      52420   50645.33
AR       Arkansas   3042017   3033946   2921964      53179   52035.48
AZ        Arizona   7640796   7520103   6407172     113990  113594.08
CA     California  39664128  39613493  37319502     163695  155779.22

or


In [116]:
states[11:16]

Name   Pop2022   Pop2021   Pop2010  TotalArea  LandArea
State                                                             
IA         Iowa   3174426   3167974   3050745      56273  55857.13
ID        Idaho   1896652   1860123   1570746      83569  82643.12
IL     Illinois  12518071  12569321  12840503      57914  55518.93
IN      Indiana   6842385   6805663   6490432      36420  35826.11
KS       Kansas   2919179   2917224   2858190      82278  81758.72

or using these index values:

In [117]:
states['IA': 'KS']

Name   Pop2022   Pop2021   Pop2010  TotalArea  LandArea
State                                                             
IA         Iowa   3174426   3167974   3050745      56273  55857.13
ID        Idaho   1896652   1860123   1570746      83569  82643.12
IL     Illinois  12518071  12569321  12840503      57914  55518.93
IN      Indiana   6842385   6805663   6490432      36420  35826.11
KS       Kansas   2919179   2917224   2858190      82278  81758.72


![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/divider.png)

## Retrieving particular columns
We can get a column of a DataFrame by using the column name:

In [118]:
states[['Pop2022']]

Pop2022
State          
AK       720763
AL      4949697
AR      3042017
AZ      7640796
CA     39664128
CO      5961083
CT      3546588
DE       998619
FL     22177997
GA     10936299
HI      1401709
IA      3174426
ID      1896652
IL     12518071
IN      6842385
KS      2919179
KY      4487233
LA      4616106
MA      6922107
MD      6075314
ME      1359677
MI      9995212
MN      5739781
MO      6184843
MS      2961536
MT      1093117
NC     10807491
ND       774008
NE      1960790
NH      1378449
NJ      8870685
NM      2109093
NV      3238601
NY     19223191
OH     11727377
OK      4007179
OR      4325290
PA     12805190
RI      1062583
SC      5342388
SD       902542
TN      7001803
TX     30097526
UT      3363182
VA      8638218
VT       622882
WA      7887965
WI      5867518
WV      1755715
WY       582233

and we can get multiple columns by passing a list of column names

In [119]:
states[['Pop2022', 'Pop2010']]

Pop2022   Pop2010
State                    
AK       720763    713910
AL      4949697   4785437
AR      3042017   2921964
AZ      7640796   6407172
CA     39664128  37319502
CO      5961083   5047349
CT      3546588   3579114
DE       998619    899593
FL     22177997  18845537
GA     10936299   9711881
HI      1401709   1363963
IA      3174426   3050745
ID      1896652   1570746
IL     12518071  12840503
IN      6842385   6490432
KS      2919179   2858190
KY      4487233   4348181
LA      4616106   4544532
MA      6922107   6566307
MD      6075314   5788645
ME      1359677   1327629
MI      9995212   9877510
MN      5739781   5310828
MO      6184843   5995974
MS      2961536   2970548
MT      1093117    990697
NC     10807491   9574323
ND       774008    674715
NE      1960790   1829542
NH      1378449   1316762
NJ      8870685   8799446
NM      2109093   2064552
NV      3238601   2702405
NY     19223191  19399878
OH     11727377  11539336
OK      4007179   3759944
OR      4325290   3837491
PA     12805190  12711160
RI      1062583   1053959
SC      5342388   4635649
SD       902542    816166
TN      7001803   6355311
TX     30097526  25241971
UT      3363182   2775332
VA      8638218   8023699
VT       622882    625879
WA      7887965   6742830
WI      5867518   5690475
WV      1755715   1854239
WY       582233    564487

### Returning rows with loc

As we have seen we can retrieve rows using the loc function. 

We can also get rows that match a specific criterion. For example, all states whose 2022 population exceeded 10 million. 

In [120]:
populousStates = states.loc[states['Pop2022'] > 10000000] 
populousStates

Name   Pop2022   Pop2021   Pop2010  TotalArea   LandArea
State                                                                    
CA         California  39664128  39613493  37319502     163695  155779.22
FL            Florida  22177997  21944577  18845537      65758   53624.76
GA            Georgia  10936299  10830007   9711881      59425   57513.49
IL           Illinois  12518071  12569321  12840503      57914   55518.93
NC     North Carolina  10807491  10701022   9574323      53819   48617.91
NY           New York  19223191  19299981  19399878      54555   47126.40
OH               Ohio  11727377  11714618  11539336      44826   40860.69
PA       Pennsylvania  12805190  12804123  12711160      46054   44742.70
TX              Texas  30097526  29730311  25241971     268596  261231.71

Or let's say we are interested in the states that lost population between 2021 and 2022.

In [121]:
states.loc[states['Pop2022'] < states['Pop2021']]

Name   Pop2022   Pop2021   Pop2010  TotalArea   LandArea
State                                                                   
AK            Alaska    720763    724357    713910     665384  570640.95
CT       Connecticut   3546588   3552821   3579114       5543    4842.36
HI            Hawaii   1401709   1406430   1363963      10932    6422.63
IL          Illinois  12518071  12569321  12840503      57914   55518.93
LA         Louisiana   4616106   4627002   4544532      52378   43293.90
MS       Mississippi   2961536   2966407   2970548      48432   43923.27
NJ        New Jersey   8870685   8874520   8799446       8723    7354.22
NY          New York  19223191  19299981  19399878      54555   47126.40
VT           Vermont    622882    623251    625879       9616    9216.66
WV     West Virginia   1755715   1767859   1854239      24230   24038.21

States that are over 10 million population and have lost population in 2022:

In [122]:
states.loc[(states['Pop2022'] < states['Pop2021']) & (states['Pop2022'] > 10000000)]

Name   Pop2022   Pop2021   Pop2010  TotalArea  LandArea
State                                                             
IL     Illinois  12518071  12569321  12840503      57914  55518.93
NY     New York  19223191  19299981  19399878      54555  47126.40


![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/divider.png)

## Creating new columns

#### State Density

![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/stateDensity.png)

Let's say we want to add a new column, Density, that gives the number of people per square mile of land area. So the Formula is



$$Density=\frac{Pop2022}{LandArea}$$

We can do that with


In [123]:
states['Density'] = states['Pop2022'] / states['LandArea']
states[:5][['Name', 'Density']]

Name     Density
State                        
AK         Alaska    1.263076
AL        Alabama   97.732545
AR       Arkansas   58.460439
AZ        Arizona   67.264033
CA     California  254.617580


![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/divider.png)

## Sorting

Density is nice but it would be nice if we ordered the DataFrame by Density...

In [124]:
states.sort_values(by=['Density'])

Name   Pop2022   Pop2021   Pop2010  TotalArea   LandArea  \
State                                                                       
AK             Alaska    720763    724357    713910     665384  570640.95   
WY            Wyoming    582233    581075    564487      97813   97093.14   
MT            Montana   1093117   1085004    990697     147040  145545.80   
ND       North Dakota    774008    770026    674715      70698   69000.80   
SD       South Dakota    902542    896581    816166      77116   75811.00   
NM         New Mexico   2109093   2105005   2064552     121590  121298.15   
ID              Idaho   1896652   1860123   1570746      83569   82643.12   
NE           Nebraska   1960790   1951996   1829542      77348   76824.17   
NV             Nevada   3238601   3185786   2702405     110572  109781.18   
KS             Kansas   2919179   2917224   2858190      82278   81758.72   
UT               Utah   3363182   3310774   2775332      84897   82169.62   
ME              Maine   1359677   1354522   1327629      35380   30842.92   
OR             Oregon   4325290   4289439   3837491      98379   95988.01   
IA               Iowa   3174426   3167974   3050745      56273   55857.13   
CO           Colorado   5961083   5893634   5047349     104094  103641.89   
OK           Oklahoma   4007179   3990443   3759944      69899   68594.92   
AR           Arkansas   3042017   3033946   2921964      53179   52035.48   
AZ            Arizona   7640796   7520103   6407172     113990  113594.08   
MS        Mississippi   2961536   2966407   2970548      48432   43923.27   
VT            Vermont    622882    623251    625879       9616    9216.66   
MN          Minnesota   5739781   5706398   5310828      86936   79626.74   
WV      West Virginia   1755715   1767859   1854239      24230   24038.21   
MO           Missouri   6184843   6169038   5995974      69707   68741.52   
AL            Alabama   4949697   4934193   4785437      52420   50645.33   
LA          Louisiana   4616106   4627002   4544532      52378   43293.90   
WI          Wisconsin   5867518   5852490   5690475      65496   54157.80   
KY           Kentucky   4487233   4480713   4348181      40408   39486.34   
TX              Texas  30097526  29730311  25241971     268596  261231.71   
WA         Washington   7887965   7796941   6742830      71298   66455.52   
NH      New Hampshire   1378449   1372203   1316762       9349    8952.65   
TN          Tennessee   7001803   6944260   6355311      42114   41234.90   
MI           Michigan   9995212   9992427   9877510      96714   56538.90   
SC     South Carolina   5342388   5277830   4635649      32020   30060.70   
GA            Georgia  10936299  10830007   9711881      59425   57513.49   
IN            Indiana   6842385   6805663   6490432      36420   35826.11   
HI             Hawaii   1401709   1406430   1363963      10932    6422.63   
VA           Virginia   8638218   8603985   8023699      42775   39490.09   
NC     North Carolina  10807491  10701022   9574323      53819   48617.91   
IL           Illinois  12518071  12569321  12840503      57914   55518.93   
CA         California  39664128  39613493  37319502     163695  155779.22   
PA       Pennsylvania  12805190  12804123  12711160      46054   44742.70   
OH               Ohio  11727377  11714618  11539336      44826   40860.69   
NY           New York  19223191  19299981  19399878      54555   47126.40   
FL            Florida  22177997  21944577  18845537      65758   53624.76   
DE           Delaware    998619    990334    899593       2489    1948.54   
MD           Maryland   6075314   6065436   5788645      12406    9707.24   
CT        Connecticut   3546588   3552821   3579114       5543    4842.36   
MA      Massachusetts   6922107   6912239   6566307      10554    7800.06   
RI       Rhode Island   1062583   1061509   1053959       1545    1033.81   
NJ         New Jersey   8870685   8874520   8799446       8723    7354.22   

           D

That is better, but perhaps we would like to order by densest states first and only show the top five ...
 

In [125]:
states.sort_values(by=['Density'], ascending=False)[:5]

Name  Pop2022  Pop2021  Pop2010  TotalArea  LandArea  \
State                                                                  
NJ        New Jersey  8870685  8874520  8799446       8723   7354.22   
RI      Rhode Island  1062583  1061509  1053959       1545   1033.81   
MA     Massachusetts  6922107  6912239  6566307      10554   7800.06   
CT       Connecticut  3546588  3552821  3579114       5543   4842.36   
MD          Maryland  6075314  6065436  5788645      12406   9707.24   

           Density  
State               
NJ     1206.203377  
RI     1027.832000  
MA      887.442789  
CT      732.408991  
MD      625.853899

Finally, let's add a column, growth, that shows the percent change in population from 2021 to 2022.  That formula would be

$$Growth=(\frac{Pop2022}{Pop2021} - 1) \times 100  $$

In [126]:
states['Growth'] = (states['Pop2022'] / states['Pop2021'] - 1) * 100
states[:10]

Name   Pop2022   Pop2021   Pop2010  TotalArea   LandArea  \
State                                                                    
AK          Alaska    720763    724357    713910     665384  570640.95   
AL         Alabama   4949697   4934193   4785437      52420   50645.33   
AR        Arkansas   3042017   3033946   2921964      53179   52035.48   
AZ         Arizona   7640796   7520103   6407172     113990  113594.08   
CA      California  39664128  39613493  37319502     163695  155779.22   
CO        Colorado   5961083   5893634   5047349     104094  103641.89   
CT     Connecticut   3546588   3552821   3579114       5543    4842.36   
DE        Delaware    998619    990334    899593       2489    1948.54   
FL         Florida  22177997  21944577  18845537      65758   53624.76   
GA         Georgia  10936299  10830007   9711881      59425   57513.49   

          Density    Growth  
State                        
AK       1.263076 -0.496164  
AL      97.732545  0.314216  
AR      58.460439  0.266023  
AZ      67.264033  1.604938  
CA     254.617580  0.127823  
CO      57.516155  1.144438  
CT     732.408991 -0.175438  
DE     512.496023  0.836586  
FL     413.577553  1.063680  
GA     190.151893  0.981458

And let's find the 5 fastest growing states:

In [127]:
states.sort_values(by=['Growth'], ascending=False)[:5]

Name   Pop2022   Pop2021   Pop2010  TotalArea   LandArea  \
State                                                                
ID       Idaho   1896652   1860123   1570746      83569   82643.12   
NV      Nevada   3238601   3185786   2702405     110572  109781.18   
AZ     Arizona   7640796   7520103   6407172     113990  113594.08   
UT        Utah   3363182   3310774   2775332      84897   82169.62   
TX       Texas  30097526  29730311  25241971     268596  261231.71   

          Density    Growth  
State                        
ID      22.949908  1.963795  
NV      29.500512  1.657833  
AZ      67.264033  1.604938  
UT      40.929750  1.582953  
TX     115.213907  1.235154


![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/divider.png)

## Descriptive Statistics on DataFrames
One handy function is `describe`

In [128]:
states.describe()

Pop2022       Pop2021       Pop2010     TotalArea       LandArea  \
count  5.000000e+01  5.000000e+01  5.000000e+01      50.00000      50.000000   
mean   6.643545e+06  6.612588e+06  6.174329e+06   75932.92000   70578.687800   
std    7.492630e+06  7.457727e+06  6.862379e+06   97327.62071   85833.058576   
min    5.822330e+05  5.810750e+05  5.644870e+05    1545.00000    1033.810000   
25%    1.912686e+06  1.883091e+06  1.835716e+06   37417.00000   36741.167500   
50%    4.551670e+06  4.553858e+06  4.446356e+06   57093.50000   53891.280000   
75%    7.826173e+06  7.727732e+06  6.698699e+06   84565.00000   81225.725000   
max    3.966413e+07  3.961349e+07  3.731950e+07  665384.00000  570640.950000   

           Density     Growth  
count    50.000000  50.000000  
mean    204.271834   0.437188  
std     266.566750   0.599162  
min       1.263076  -0.686933  
25%      48.003341   0.075556  
50%     107.481847   0.290119  
75%     221.406815   0.834008  
max    1206.203377   1.963795

Alternatively, we could retrieve a specific statistic:


In [129]:
states['Pop2022'].mean()

6643544.68

or find out how many states lost population in 2022 using `count`:

In [130]:
states.loc[states['Pop2022'] < states['Pop2021']][['Pop2022']].count()

Pop2022    10
dtype: int64

Or get the total population of the United States (minus the District of Columbia)

In [131]:
states['Pop2022'].sum()

332177234

or to make that more readable

In [132]:
totalPop = "{:,}".format(states['Pop2022'].sum())
print('Total Population: ' + totalPop)

Total Population: 332,177,234


What percentage of the U. S. population is accounted for by the top 5 most populous states?

In [133]:
statesByPopulation = states.sort_values(by=['Pop2022'], ascending=False)[:10][['Pop2022']]
percent = (statesByPopulation.sum() / states['Pop2022'].sum()) * 100
print("The top 10 most populous states account for {:4.2f} percent of the U.S. Population" .format(percent['Pop2022']))

The top 10 most populous states account for 54.17 percent of the U.S. Population


## Summary Statistics and Axes

![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/ax.png)
Suppose I have customers of my vinyl record shop rate different artists

|Customer | Taylor Swift | Miranda Lambert | Carrie Underwood | Nicki Minaj | Ariana Grande |
|:-----------|:------:|:------:|:---------:|:------:|:--------:|
|Jake|5|-|5|2|2|
|Clara|2|-|-|4|5|
|Kelsey|5|5|5|2|-|
|Angelica|2|3|-|5|5|
|Jordyn|2|1|-|5|-|

First we will read in the file ...

In [134]:
ratings = pd.read_csv('https://raw.githubusercontent.com/zacharski/ml-class/master/data/ratings.csv', index_col=0)
ratings

Taylor Swift  Miranda Lambert  Carrie Underwood  Nicki Minaj  \
Customer                                                                 
Jake                 5              NaN               5.0            2   
Clara                2              NaN               NaN            4   
Kelsey               5              5.0               5.0            2   
Angelica             2              3.0               NaN            5   
Jordyn               2              1.0               NaN            5   

          Ariana Grande  
Customer                 
Jake                2.0  
Clara               5.0  
Kelsey              NaN  
Angelica            5.0  
Jordyn              NaN

We can get the mean rating of each artist by:


In [135]:
ratings.mean()

Taylor Swift        3.2
Miranda Lambert     3.0
Carrie Underwood    5.0
Nicki Minaj         3.6
Ariana Grande       4.0
dtype: float64

Note that the summary statistics ignore NaN entries. The mean rating for Miranda Lambert is computed just on the people that rated her.

Many descriptive statistics functions take an optional parameter `axis` that tells which axis to reduce over. If we want the mean ratings for each **customer** instead of each artist we can do:

In [136]:
ratings.mean(axis=1)

Customer
Jake        3.500000
Clara       3.666667
Kelsey      4.250000
Angelica    3.750000
Jordyn      2.666667
dtype: float64



Sweet! So `axis=1` means reduce by rows and `axis=0` means reduce by columns:

In [137]:
ratings.mean(axis=0)

Taylor Swift        3.2
Miranda Lambert     3.0
Carrie Underwood    5.0
Nicki Minaj         3.6
Ariana Grande       4.0
dtype: float64

![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/divider.png)

## List of Descriptive Statistics
(from the book *Python for Data Analysis*)

Method | Description
:-- | :--
`count` | Number of non-NaN values
`describe` | A set of common summary statistics
`min, max` | compute minimum and maximum values
`argmin, argmax` | compute index locations of minimum and maximum values
`sum` | Sum the values
`mean` | Mean of values
`median` | Median of values
`std` | Sample standard deviation

So, for example, the lowest rating for each artist:


In [138]:
ratings.min()

Taylor Swift        2.0
Miranda Lambert     1.0
Carrie Underwood    5.0
Nicki Minaj         2.0
Ariana Grande       2.0
dtype: float64


![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/divider.png)

# <font color='#EE4C2C'>You Try ...</font> 
Ok, it is time for you to try out what you just learned. Let us start with the electric vehicle datafile we have already seen:

```
https://raw.githubusercontent.com/zacharski/datamining-guide/main/data/ElectricCarData_Clean.csv
```
When we loaded the datafile before, it worked fine for those examples, but you may encounter an error here, that will require you to modify the `read_csv`.

![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/evBanner.png)

In [146]:
# TO DO
!curl https://raw.githubusercontent.com/zacharski/datamining-guide/main/data/ElectricCarData_Clean.csv > tryit.csv

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  8268  100  8268    0     0   218k      0 --:--:-- --:--:-- --:--:--  218k


In [140]:
vals = pd.read_csv('tryit.csv')
vals

Brand                            Model  AccelSec  TopSpeed_KmH  \
0         Tesla    Model 3 Long Range Dual Motor       4.6           233   
1    Volkswagen                        ID.3 Pure      10.0           160   
2      Polestar                                2       4.7           210   
3           BMW                              iX3       6.8           180   
4         Honda                                e       9.5           145   
..          ...                              ...       ...           ...   
101      Nissan  Ariya e-4ORCE 87kWh Performance       5.1           200   
102       Byton                M-Byte 95 kWh 2WD       7.5           190   
103      Rivian                              R1T       3.0           185   
104         GMC                        Hummer EV       3.0           171   
105        Ford                   F150 Lightning       3.0           180   

     Range_Km  Efficiency_WhKm FastCharge_KmH RapidCharge PowerTrain  \
0         450              161            940         Yes        AWD   
1         270              167            250         Yes        RWD   
2         400              181            620         Yes        AWD   
3         360              206            560         Yes        RWD   
4         170              168            190         Yes        RWD   
..        ...              ...            ...         ...        ...   
101       375              232            450         Yes        AWD   
102       400              238            480         Yes        AWD   
103       505              299            337         Yes        AWD   
104       529              446           250          Yes        AWD   
105       514              299           309          Yes        AWD   

       PlugType  BodyStyle Segment  Seats PriceEuro  
0    Type 2 CCS      Sedan       D      5     55480  
1    Type 2 CCS  Hatchback       C      5     30000  
2    Type 2 CCS   Liftback       D      5     56440  
3    Type 2 CCS        SUV       D      5     68040  
4    Type 2 CCS  Hatchback       B      4     32997  
..          ...        ...     ...    ...       ...  
101  Type 2 CCS  Hatchback       C      5     65000  
102  Type 2 CCS        SUV       E      5     62000  
103  Type 2 CCS     Pickup       N      6     75000  
104  Type 2 CCS      Truck       E      5         -  
105  Type 2 CCS      Truck       E      5         -  

[106 rows x 14 columns]

Some of the columns are based on kilometers:

* TopSpeed_KmH	
* Range_Km	
* Efficiency_WhKm	
* FastCharge_KmH

and we are first going to add USA mile-centric columns.

## <font color='#EE4C2C'>1. TopSpeed_MPH</font>

Add a new column `TopSpeed_MPH` computed from the `TopSpeed_KmH` column, and just to check, display the first 10 rows. (The Lucid Air should have a top speed slightly over 155 MPH)


In [147]:
vals['TopSpeed_MPH'] = vals['TopSpeed_KmH'] * 0.621371
vals[:10]

Brand                          Model  AccelSec  TopSpeed_KmH  \
0       Tesla  Model 3 Long Range Dual Motor       4.6           233   
1  Volkswagen                      ID.3 Pure      10.0           160   
2    Polestar                              2       4.7           210   
3         BMW                            iX3       6.8           180   
4       Honda                              e       9.5           145   
5       Lucid                            Air       2.8           250   
6  Volkswagen                         e-Golf       9.6           150   
7     Peugeot                          e-208       8.1           150   
8       Tesla    Model 3 Standard Range Plus       5.6           225   
9        Audi                      Q4 e-tron       6.3           180   

   Range_Km  Efficiency_WhKm  FastCharge_KmH RapidCharge PowerTrain  \
0       450              161           940.0         Yes        AWD   
1       270              167           250.0         Yes        RWD   
2       400              181           620.0         Yes        AWD   
3       360              206           560.0         Yes        RWD   
4       170              168           190.0         Yes        RWD   
5       610              180           620.0         Yes        AWD   
6       190              168           220.0         Yes        FWD   
7       275              164           420.0         Yes        FWD   
8       310              153           650.0         Yes        RWD   
9       400              193           540.0         Yes        AWD   

     PlugType  BodyStyle Segment  Seats  PriceEuro  FastCharge_MPH  \
0  Type 2 CCS      Sedan       D      5    55480.0       584.08874   
1  Type 2 CCS  Hatchback       C      5    30000.0       155.34275   
2  Type 2 CCS   Liftback       D      5    56440.0       385.25002   
3  Type 2 CCS        SUV       D      5    68040.0       347.96776   
4  Type 2 CCS  Hatchback       B      4    32997.0       118.06049   
5  Type 2 CCS      Sedan       F      5   105000.0       385.25002   
6  Type 2 CCS  Hatchback       C      5    31900.0       136.70162   
7  Type 2 CCS  Hatchback       B      5    29682.0       260.97582   
8  Type 2 CCS      Sedan       D      5    46380.0       403.89115   
9  Type 2 CCS        SUV       D      5    55000.0       335.54034   

   Efficiency_MkwH  TopSpeed_MPH  
0         3.859447    144.779443  
1         3.720784     99.419360  
2         3.432989    130.487910  
3         3.016364    111.846780  
4         3.698637     90.098795  
5         3.452061    155.342750  
6         3.698637     93.205650  
7         3.788848     93.205650  
8         4.061248    139.808475  
9         3.219539    111.846780


## <font color='#EE4C2C'>2. Range and FastCharge_MPH</font>

Add a new column `Range` computed from the `Range_Km` column that shows the range in miles.

Also add a column `FastCharge_MPH computed from `FastCharge_KmH`. This will indicate how many miles can you go on one hour of charging. 

Just to check, display the first 10 rows. (The Tesla Model 3 Long Range Dual Motor should have a range of nearly 280 miles and charge at a rate of 584 miles per hour of fast charging.)

In [149]:
vals['Range'] = vals['Range_Km'] * 0.621371
vals[:10]

Brand                          Model  AccelSec  TopSpeed_KmH  \
0       Tesla  Model 3 Long Range Dual Motor       4.6           233   
1  Volkswagen                      ID.3 Pure      10.0           160   
2    Polestar                              2       4.7           210   
3         BMW                            iX3       6.8           180   
4       Honda                              e       9.5           145   
5       Lucid                            Air       2.8           250   
6  Volkswagen                         e-Golf       9.6           150   
7     Peugeot                          e-208       8.1           150   
8       Tesla    Model 3 Standard Range Plus       5.6           225   
9        Audi                      Q4 e-tron       6.3           180   

   Range_Km  Efficiency_WhKm  FastCharge_KmH RapidCharge PowerTrain  \
0       450              161           940.0         Yes        AWD   
1       270              167           250.0         Yes        RWD   
2       400              181           620.0         Yes        AWD   
3       360              206           560.0         Yes        RWD   
4       170              168           190.0         Yes        RWD   
5       610              180           620.0         Yes        AWD   
6       190              168           220.0         Yes        FWD   
7       275              164           420.0         Yes        FWD   
8       310              153           650.0         Yes        RWD   
9       400              193           540.0         Yes        AWD   

     PlugType  BodyStyle Segment  Seats  PriceEuro  FastCharge_MPH  \
0  Type 2 CCS      Sedan       D      5    55480.0       584.08874   
1  Type 2 CCS  Hatchback       C      5    30000.0       155.34275   
2  Type 2 CCS   Liftback       D      5    56440.0       385.25002   
3  Type 2 CCS        SUV       D      5    68040.0       347.96776   
4  Type 2 CCS  Hatchback       B      4    32997.0       118.06049   
5  Type 2 CCS      Sedan       F      5   105000.0       385.25002   
6  Type 2 CCS  Hatchback       C      5    31900.0       136.70162   
7  Type 2 CCS  Hatchback       B      5    29682.0       260.97582   
8  Type 2 CCS      Sedan       D      5    46380.0       403.89115   
9  Type 2 CCS        SUV       D      5    55000.0       335.54034   

   Efficiency_MkwH  TopSpeed_MPH       Range  
0         3.859447    144.779443  279.616950  
1         3.720784     99.419360  167.770170  
2         3.432989    130.487910  248.548400  
3         3.016364    111.846780  223.693560  
4         3.698637     90.098795  105.633070  
5         3.452061    155.342750  379.036310  
6         3.698637     93.205650  118.060490  
7         3.788848     93.205650  170.877025  
8         4.061248    139.808475  192.625010  
9         3.219539    111.846780  248.548400

In [143]:
vals = pd.read_csv('tryit.csv', na_values='-')
vals['FastCharge_MPH'] = pd.to_numeric(vals['FastCharge_KmH']) * 0.621371
vals[:10]

Brand                          Model  AccelSec  TopSpeed_KmH  \
0       Tesla  Model 3 Long Range Dual Motor       4.6           233   
1  Volkswagen                      ID.3 Pure      10.0           160   
2    Polestar                              2       4.7           210   
3         BMW                            iX3       6.8           180   
4       Honda                              e       9.5           145   
5       Lucid                            Air       2.8           250   
6  Volkswagen                         e-Golf       9.6           150   
7     Peugeot                          e-208       8.1           150   
8       Tesla    Model 3 Standard Range Plus       5.6           225   
9        Audi                      Q4 e-tron       6.3           180   

   Range_Km  Efficiency_WhKm  FastCharge_KmH RapidCharge PowerTrain  \
0       450              161           940.0         Yes        AWD   
1       270              167           250.0         Yes        RWD   
2       400              181           620.0         Yes        AWD   
3       360              206           560.0         Yes        RWD   
4       170              168           190.0         Yes        RWD   
5       610              180           620.0         Yes        AWD   
6       190              168           220.0         Yes        FWD   
7       275              164           420.0         Yes        FWD   
8       310              153           650.0         Yes        RWD   
9       400              193           540.0         Yes        AWD   

     PlugType  BodyStyle Segment  Seats  PriceEuro  FastCharge_MPH  
0  Type 2 CCS      Sedan       D      5    55480.0       584.08874  
1  Type 2 CCS  Hatchback       C      5    30000.0       155.34275  
2  Type 2 CCS   Liftback       D      5    56440.0       385.25002  
3  Type 2 CCS        SUV       D      5    68040.0       347.96776  
4  Type 2 CCS  Hatchback       B      4    32997.0       118.06049  
5  Type 2 CCS      Sedan       F      5   105000.0       385.25002  
6  Type 2 CCS  Hatchback       C      5    31900.0       136.70162  
7  Type 2 CCS  Hatchback       B      5    29682.0       260.97582  
8  Type 2 CCS      Sedan       D      5    46380.0       403.89115  
9  Type 2 CCS        SUV       D      5    55000.0       335.54034

## <font color='#EE4C2C'>3. Efficiency_MkwH</font>

The DataFrame has a column `Efficiency_WhKm` which indicates how many watt hours does it take to go one kilometer. One common measure of efficieny is how many miles can you go on one kilowatt hour (kwH). So please add this column to the DataFrame. (The Tesla Model 3 Standard Range Plus should have an efficiency of around 4 miles per kilowatt hour.) 


In [144]:
vals['Efficiency_MkwH'] = 1000 / vals['Efficiency_WhKm']
vals['Efficiency_MkwH'] = vals['Efficiency_MkwH'] * 0.621371
vals[:10]

Brand                          Model  AccelSec  TopSpeed_KmH  \
0       Tesla  Model 3 Long Range Dual Motor       4.6           233   
1  Volkswagen                      ID.3 Pure      10.0           160   
2    Polestar                              2       4.7           210   
3         BMW                            iX3       6.8           180   
4       Honda                              e       9.5           145   
5       Lucid                            Air       2.8           250   
6  Volkswagen                         e-Golf       9.6           150   
7     Peugeot                          e-208       8.1           150   
8       Tesla    Model 3 Standard Range Plus       5.6           225   
9        Audi                      Q4 e-tron       6.3           180   

   Range_Km  Efficiency_WhKm  FastCharge_KmH RapidCharge PowerTrain  \
0       450              161           940.0         Yes        AWD   
1       270              167           250.0         Yes        RWD   
2       400              181           620.0         Yes        AWD   
3       360              206           560.0         Yes        RWD   
4       170              168           190.0         Yes        RWD   
5       610              180           620.0         Yes        AWD   
6       190              168           220.0         Yes        FWD   
7       275              164           420.0         Yes        FWD   
8       310              153           650.0         Yes        RWD   
9       400              193           540.0         Yes        AWD   

     PlugType  BodyStyle Segment  Seats  PriceEuro  FastCharge_MPH  \
0  Type 2 CCS      Sedan       D      5    55480.0       584.08874   
1  Type 2 CCS  Hatchback       C      5    30000.0       155.34275   
2  Type 2 CCS   Liftback       D      5    56440.0       385.25002   
3  Type 2 CCS        SUV       D      5    68040.0       347.96776   
4  Type 2 CCS  Hatchback       B      4    32997.0       118.06049   
5  Type 2 CCS      Sedan       F      5   105000.0       385.25002   
6  Type 2 CCS  Hatchback       C      5    31900.0       136.70162   
7  Type 2 CCS  Hatchback       B      5    29682.0       260.97582   
8  Type 2 CCS      Sedan       D      5    46380.0       403.89115   
9  Type 2 CCS        SUV       D      5    55000.0       335.54034   

   Efficiency_MkwH  
0         3.859447  
1         3.720784  
2         3.432989  
3         3.016364  
4         3.698637  
5         3.452061  
6         3.698637  
7         3.788848  
8         4.061248  
9         3.219539

## <font color='#EE4C2C'>4. New DataFrame</font>

Create a new DataFrame, `bevs` (battery electric vehicles as opposed to PHEVs) that contain only the following columns from the current DataFrame

* Brand	
* Model	
* PowerTrain	
* AccelSec
* TopSpeed_MPH	
* Range	
* Efficiency_MkwH

```
'Brand', 'Model', 'PowerTrain', 'AccelSec', 'TopSpeed_MPH', 'Range', 'Efficiency_MkwH'
```


In [150]:
bevs = pd.DataFrame({'Brand': vals['Brand'], 'Model': vals['Model'], 'PowerTrain': vals['PowerTrain'], 'AccelSec': vals['AccelSec'], 'TopSpeed_MPH': vals['TopSpeed_MPH'], 'Range': vals['Range'], 'Efficiency_MkwH': vals['Efficiency_MkwH']})
bevs[:10]

Brand                          Model PowerTrain  AccelSec  \
0       Tesla  Model 3 Long Range Dual Motor        AWD       4.6   
1  Volkswagen                      ID.3 Pure        RWD      10.0   
2    Polestar                              2        AWD       4.7   
3         BMW                            iX3        RWD       6.8   
4       Honda                              e        RWD       9.5   
5       Lucid                            Air        AWD       2.8   
6  Volkswagen                         e-Golf        FWD       9.6   
7     Peugeot                          e-208        FWD       8.1   
8       Tesla    Model 3 Standard Range Plus        RWD       5.6   
9        Audi                      Q4 e-tron        AWD       6.3   

   TopSpeed_MPH       Range  Efficiency_MkwH  
0    144.779443  279.616950         3.859447  
1     99.419360  167.770170         3.720784  
2    130.487910  248.548400         3.432989  
3    111.846780  223.693560         3.016364  
4     90.098795  105.633070         3.698637  
5    155.342750  379.036310         3.452061  
6     93.205650  118.060490         3.698637  
7     93.205650  170.877025         3.788848  
8    139.808475  192.625010         4.061248  
9    111.846780  248.548400         3.219539

## <font color='#EE4C2C'>5. Most and least efficient</font>
Using the bevs DataFrame, what are the 5 most efficient evs ordered by the most efficient first?


In [151]:
bevs.sort_values(by=['Efficiency_MkwH'], ascending = False)[:5]

Brand                        Model PowerTrain  AccelSec  TopSpeed_MPH  \
48  Lightyear                          One        AWD      10.0     93.205650   
8       Tesla  Model 3 Standard Range Plus        RWD       5.6    139.808475   
14    Hyundai               IONIQ Electric        FWD       9.7    102.526215   
41    Hyundai         Kona Electric 39 kWh        FWD       9.9     96.312505   
19       Mini                    Cooper SE        FWD       7.3     93.205650   

         Range  Efficiency_MkwH  
48  357.288325         5.974721  
8   192.625010         4.061248  
14  155.342750         4.061248  
41  158.449605         4.034877  
19  114.953635         3.983147

What are the five most inefficient evs ordered by least efficient first?

In [152]:
bevs.sort_values(by=['Efficiency_MkwH'])[:5]

Brand                Model PowerTrain  AccelSec  TopSpeed_MPH  \
104       GMC            Hummer EV        AWD       3.0    106.254441   
105      Ford       F150 Lightning        AWD       3.0    111.846780   
103    Rivian                  R1T        AWD       3.0    114.953635   
84   Mercedes         EQV 300 Long        FWD      10.0     86.991940   
90       Audi  e-tron S 55 quattro        AWD       4.5    130.487910   

          Range  Efficiency_MkwH  
104  328.705259         1.393209  
105  319.384694         2.078164  
103  313.792355         2.078164  
84   205.052430         2.276084  
90   198.838720         2.301374

## <font color='#EE4C2C'>6. fastest
![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/porsche.png)

The Porsche 911 997 Turbo S is among the world's fastest production combustion cars with a 0-60 acceleration of 2.9 seconds. (In our dataset `AccelSec` measures the same thing.) Which electric cars have better acceleration than this? 

In [153]:
bevs2 = bevs.loc[bevs['AccelSec'] < 3.0]
bevs2

Brand                Model PowerTrain  AccelSec  TopSpeed_MPH  \
5     Lucid                  Air        AWD       2.8    155.342750   
16  Porsche       Taycan Turbo S        AWD       2.8    161.556460   
51    Tesla             Roadster        AWD       2.1    254.762110   
59    Tesla  Model S Performance        AWD       2.5    162.177831   
81    Tesla  Model X Performance        AWD       2.8    155.342750   

         Range  Efficiency_MkwH  
5   379.036310         3.452061  
16  233.014125         2.786417  
51  602.729870         3.016364  
59  313.792355         3.305165  
81  273.403240         2.876718

## <font color='#EE4C2C'>7. Refrigerators
![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/frigTesla.png)

Representative Thomas Massie (R-KY) at a hearing questioning Secretarty Pete Buttigieg says the energy (kWh) required to home charge an electric vehicle for a year is 25 times greater than running the average refrigerator. He also said "It would take four times as much electricity to charge the average household's cars as the average household uses on air conditioning." And he said, "that means the average household would use twice as much electricity charging one of their cars as they would use for all of the air conditioning that they use for the entire year." Is this true? 

Since air conditioner use is quite varied in different regions of the country, perhaps a better measure would be the energy used for air conditioning and heating combined.

You are going to need to do some web searching to find the typical energy use of household these household items.

What do you think about the usefulness of this comparing household appliances to EVs when making government policy?  


In [154]:
!curl https://raw.githubusercontent.com/zacharski/datamining-guide/main/data/IEA-EV-data.csv > '8.csv'

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  487k  100  487k    0     0  1517k      0 --:--:-- --:--:-- --:--:-- 1517k


In [155]:
Fridge = pd.read_csv('8.csv')
Fridge

region          category              parameter    mode powertrain  \
0     Australia        Historical               EV sales    Cars        BEV   
1     Australia        Historical               EV stock    Cars        BEV   
2     Australia        Historical         EV sales share    Cars         EV   
3     Australia        Historical         EV stock share    Cars         EV   
4     Australia        Historical         EV stock share    Cars         EV   
...         ...               ...                    ...     ...        ...   
7011      World  Projection-STEPS   Oil displacement Mbd    Cars         EV   
7012      World  Projection-STEPS  Oil displacement Mlge   Buses         EV   
7013      World  Projection-STEPS  Oil displacement Mlge  Trucks         EV   
7014      World  Projection-STEPS  Oil displacement Mlge    Vans         EV   
7015      World  Projection-STEPS  Oil displacement Mlge    Cars         EV   

      year                               unit          value  
0     2011                              sales      49.000000  
1     2011                              stock      49.000000  
2     2011                            percent       0.006479  
3     2011                            percent       0.000464  
4     2012                            percent       0.002785  
...    ...                                ...            ...  
7011  2030             Milion barrels per day       2.409347  
7012  2030  Milion litres gasoline equivalent   30311.880859  
7013  2030  Milion litres gasoline equivalent   13943.429688  
7014  2030  Milion litres gasoline equivalent   16207.837891  
7015  2030  Milion litres gasoline equivalent  141228.515625  

[7016 rows x 8 columns]

In [156]:
Fridge1 = Fridge.loc[(Fridge['powertrain'] == "EV") & (Fridge['parameter'] == "EV sales share") & (Fridge['year'] == 2021) & (Fridge['unit'] == "percent") & (Fridge['category'] == "Historical") & (Fridge['mode'] == "Cars")]
Fridge1.sort_values(by=['value'], ascending = False)[:5]

region    category       parameter  mode powertrain  year     unit  \
4490   Norway  Historical  EV sales share  Cars         EV  2021  percent   
3012  Iceland  Historical  EV sales share  Cars         EV  2021  percent   
5873   Sweden  Historical  EV sales share  Cars         EV  2021  percent   
1572  Denmark  Historical  EV sales share  Cars         EV  2021  percent   
2311  Finland  Historical  EV sales share  Cars         EV  2021  percent   

          value  
4490  86.167145  
3012  71.728737  
5873  43.294743  
1572  35.424446  
2311  30.780634

## <font color='#EE4C2C'>8. Global Adoption of EVs

![](https://raw.githubusercontent.com/zacharski/datamining-guide/master/labs/pics/earth.png)

The data file we are using for this is

```
https://raw.githubusercontent.com/zacharski/datamining-guide/main/data/IEA-EV-data.csv
```

This data is from the [International Energy Agency](https://www.iea.org/data-and-statistics/data-product/global-ev-outlook-2022).

The first step in data mining work is to examine the dataset, which often is not in the best format to directly answer our questions. That is the case with this file. You have the Pandas skills, but you will need to spend a bit of time looking at the datafile and understanding it, so you can write some code to meet the need.

So what is our need? We are interested in what percent of new car sales are evs in 2021. We are interested in historical data, not predictions. We would like to see the top five countries with the best adoption percent and the bottom five. Have appropriate labels indicating the list of top 5 and bottom 5. 



In [157]:
!curl https://raw.githubusercontent.com/zacharski/datamining-guide/main/data/IEA-EV-data.csv > '8.csv'

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  487k  100  487k    0     0  6244k      0 --:--:-- --:--:-- --:--:-- 6244k


In [158]:
GAEV = pd.read_csv('8.csv')
GAEV

region          category              parameter    mode powertrain  \
0     Australia        Historical               EV sales    Cars        BEV   
1     Australia        Historical               EV stock    Cars        BEV   
2     Australia        Historical         EV sales share    Cars         EV   
3     Australia        Historical         EV stock share    Cars         EV   
4     Australia        Historical         EV stock share    Cars         EV   
...         ...               ...                    ...     ...        ...   
7011      World  Projection-STEPS   Oil displacement Mbd    Cars         EV   
7012      World  Projection-STEPS  Oil displacement Mlge   Buses         EV   
7013      World  Projection-STEPS  Oil displacement Mlge  Trucks         EV   
7014      World  Projection-STEPS  Oil displacement Mlge    Vans         EV   
7015      World  Projection-STEPS  Oil displacement Mlge    Cars         EV   

      year                               unit          value  
0     2011                              sales      49.000000  
1     2011                              stock      49.000000  
2     2011                            percent       0.006479  
3     2011                            percent       0.000464  
4     2012                            percent       0.002785  
...    ...                                ...            ...  
7011  2030             Milion barrels per day       2.409347  
7012  2030  Milion litres gasoline equivalent   30311.880859  
7013  2030  Milion litres gasoline equivalent   13943.429688  
7014  2030  Milion litres gasoline equivalent   16207.837891  
7015  2030  Milion litres gasoline equivalent  141228.515625  

[7016 rows x 8 columns]

In [159]:
GAEV2 = GAEV.loc[(GAEV['powertrain'] == "EV") & (GAEV['year'] == 2021) & (GAEV['unit'] == "percent") & (GAEV['category'] == "Historical") & (GAEV['mode'] == "Cars")]
GAEV2.sort_values(by=['value'], ascending = False)[:5]

region    category       parameter  mode powertrain  year     unit  \
4490   Norway  Historical  EV sales share  Cars         EV  2021  percent   
3012  Iceland  Historical  EV sales share  Cars         EV  2021  percent   
5873   Sweden  Historical  EV sales share  Cars         EV  2021  percent   
1572  Denmark  Historical  EV sales share  Cars         EV  2021  percent   
2311  Finland  Historical  EV sales share  Cars         EV  2021  percent   

          value  
4490  86.167145  
3012  71.728737  
5873  43.294743  
1572  35.424446  
2311  30.780634

In [160]:
GAEV2 = GAEV.loc[(GAEV['powertrain'] == "EV") & (GAEV['year'] == 2021) & (GAEV['unit'] == "percent") & (GAEV['category'] == "Historical") & (GAEV['mode'] == "Cars")]
GAEV2.sort_values(by=['value'])[:5]

region    category       parameter  mode powertrain  year  \
5499  South Africa  Historical  EV stock share  Cars         EV  2021   
361         Brazil  Historical  EV stock share  Cars         EV  2021   
578          Chile  Historical  EV stock share  Cars         EV  2021   
3917        Mexico  Historical  EV stock share  Cars         EV  2021   
3138         India  Historical  EV stock share  Cars         EV  2021   

         unit     value  
5499  percent  0.025517  
361   percent  0.032904  
578   percent  0.043362  
3917  percent  0.056817  
3138  percent  0.059175

## <font color='#EE4C2C'>9. Adoption of EVs Worldwide Part 2
In the U.S. in 2021 Electric Vehicles account for 2.5% of total new car sales. Is this higher or lower than the average (mean) 2021 figures in our dataset?

Also, how many countries have a higher adoption rate than the U.S?

In addition to cells showing your work, please have some readable content.

In [161]:
GAEV3 = GAEV.loc[(GAEV['region'] == "USA") & (GAEV['powertrain'] == "EV") & (GAEV['year'] == 2021) & (GAEV['unit'] == "percent") & (GAEV['category'] == "Historical") & (GAEV['mode'] == "Cars")]
GAEV3.mean()

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  


year     2021.000000
value       2.730079
dtype: float64

In [162]:
GAEV4 = GAEV.loc[(GAEV['region'] != "USA") & (GAEV['powertrain'] == "EV") & (GAEV['year'] == 2021) & (GAEV['unit'] == "percent") & (GAEV['category'] == "Historical") & (GAEV['mode'] == "Cars")]
GAEV4

region    category       parameter  mode powertrain  year  \
66         Australia  Historical  EV stock share  Cars         EV  2021   
73         Australia  Historical  EV sales share  Cars         EV  2021   
287          Belgium  Historical  EV sales share  Cars         EV  2021   
288          Belgium  Historical  EV stock share  Cars         EV  2021   
361           Brazil  Historical  EV stock share  Cars         EV  2021   
362           Brazil  Historical  EV sales share  Cars         EV  2021   
484           Canada  Historical  EV sales share  Cars         EV  2021   
493           Canada  Historical  EV stock share  Cars         EV  2021   
578            Chile  Historical  EV stock share  Cars         EV  2021   
579            Chile  Historical  EV sales share  Cars         EV  2021   
1092           China  Historical  EV sales share  Cars         EV  2021   
1094           China  Historical  EV stock share  Cars         EV  2021   
1572         Denmark  Historical  EV sales share  Cars         EV  2021   
1573         Denmark  Historical  EV stock share  Cars         EV  2021   
1896          Europe  Historical  EV sales share  Cars         EV  2021   
1897          Europe  Historical  EV stock share  Cars         EV  2021   
2311         Finland  Historical  EV sales share  Cars         EV  2021   
2312         Finland  Historical  EV stock share  Cars         EV  2021   
2546          France  Historical  EV sales share  Cars         EV  2021   
2547          France  Historical  EV stock share  Cars         EV  2021   
2749         Germany  Historical  EV sales share  Cars         EV  2021   
2750         Germany  Historical  EV stock share  Cars         EV  2021   
2842          Greece  Historical  EV sales share  Cars         EV  2021   
2843          Greece  Historical  EV stock share  Cars         EV  2021   
3012         Iceland  Historical  EV sales share  Cars         EV  2021   
3013         Iceland  Historical  EV stock share  Cars         EV  2021   
3138           India  Historical  EV stock share  Cars         EV  2021   
3139           India  Historical  EV sales share  Cars         EV  2021   
3510           Italy  Historical  EV stock share  Cars         EV  2021   
3513           Italy  Historical  EV sales share  Cars         EV  2021   
3688           Japan  Historical  EV stock share  Cars         EV  2021   
3689           Japan  Historical  EV sales share  Cars         EV  2021   
3839           Korea  Historical  EV sales share  Cars         EV  2021   
3851           Korea  Historical  EV stock share  Cars         EV  2021   
3916          Mexico  Historical  EV sales share  Cars         EV  2021   
3917          Mexico  Historical  EV stock share  Cars         EV  2021   
4137     Netherlands  Historical  EV stock share  Cars         EV  2021   
4139     Netherlands  Historical  EV sales share  Cars         EV  2021   
4284     New Zealand  Historical  EV stock share  Cars         EV  2021   
4285     New Zealand  Historical  EV sales share  Cars         EV  2021   
4490          Norway  Historical  EV sales share  Cars         EV  2021   
4493          Norway  Historical  EV stock share  Cars         EV  2021   
4705    Other Europe  Historical  EV sales share  Cars         EV  2021   
4853          Poland  Historical  EV sales share  Cars         EV  2021   
4854          Poland  Historical  EV stock share  Cars         EV  2021   
4979        Portugal  Historical  EV stock share  Cars         EV  2021   
4992        Portugal  Historical  EV sales share  Cars         EV  2021   
5498    South Africa  Historical  EV sales share  Cars         EV  2021   
5499    South Africa  Historical  EV stock share  Cars         EV  2021   
5669           Spain  Historical  EV stock share  Cars         EV  2021   
5670           Spain  Historical  EV sales share  Cars         EV  2021   
5870          Sweden  Historical  EV stock share  Cars         EV  2021   
5873          Sweden  Historical  EV sa

In [163]:
GAEV4.mean()

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  """Entry point for launching an IPython kernel.


year     2021.000000
value      10.140427
dtype: float64